### PR TITLE
BUG: histogram weights aren't dropped if NaN values in data

### LINKED ISSUE #48884: BUG: histogram weights aren't dropped if NaN values in data
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
>>> import numpy as np
>>> from pandas import DataFrame
>>> df = DataFrame(np.random.randn(100), columns=["a"])
>>> weights = np.ones(100)
>>> df.loc[0, "a"] = np.nan
>>> df.plot.hist() # works fine
>>> df.plot.hist(weights=weights) # fails
```

### Issue Description

If a DataFrame contains NaN values, these are dropped to plot a histogram. However, if weights are provided, then the corresponding weights aren't dropped, so weights ends up longer than the plotted data:

```
Traceback (most recent call last):
  File "/Users/adam/Programming/multinesttest/pandas_issue.py", line 10, in <module>
    df.plot.hist(weights=weights) # fails
  File "/Users/adam/Programming/env_sep30/lib/python3.10/site-packages/pandas/plotting/_core.py", line 1375, in hist
    return self(kind="hist", by=by, bins=bins, **kwargs)
  File "/Users/adam/Programming/env_sep30/lib/python3.10/site-packages/pandas/plotting/_core.py", line 1001, in __call__
    return plot_backend.plot(data, kind=kind, **kwargs)
  File "/Users/adam/Programming/env_sep30/lib/python3.10/site-packages/pandas/plotting/_matplotlib/__init__.py", line 71, in plot
    plot_obj.generate()
  File "/Users/adam/Programming/env_sep30/lib/python3.10/site-packages/pandas/plotting/_matplotlib/core.py", line 453, in generate
    self._make_plot()
  File "/Users/adam/Programming/env_sep30/lib/python3.10/site-packages/pandas/plotting/_matplotlib/hist.py", line 154, in _make_plot
    artists = self._plot(ax, y, column_num=i, stacking_id=stacking_id, **kwds)
  File "/Users/adam/Programming/env_sep30/lib/python3.10/site-packages/pandas/plotting/_matplotlib/hist.py", line 108, in _plot
    n, bins, patches = ax.hist(y, bins=bins, bottom=bottom, **kwds)
  File "/Users/adam/Programming/env_sep30/lib/python3.10/site-packages/matplotlib/__init__.py", line 1423, in inner
    return func(ax, *map(sanitize_sequence, args), **kwargs)
  File "/Users/adam/Programming/env_sep30/lib/python3.10/site-packages/matplotlib/axes/_axes.py", line 6652, in hist
    raise ValueError('weights should have the same shape as x')
ValueError: weights should have the same shape as x
```

### Expected Behavior

The weights corresponding to the NaN data should be dropped before plotting the histogram.

I believe the problem lies here:

https://github.com/pandas-dev/pandas/blob/71c94c35add4fee5752cedf742ce443637585fe2/pandas/plotting/_matplotlib/hist.py#L147-L156

I think the order should be: weights is popped, drop the weights corresponding to the NaN values of y, before `y = reformat_hist_y_given_by(y, self.by)`.

### Installed Versions

<details>

Replace this line with the output of pd.show_versions()

</details>
INSTALLED VERSIONS

------------------

commit           : 87cfe4e38bafe7300a6003a1d18bd80f3f77c763
python           : 3.10.6.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 21.6.0
Version          : Darwin Kernel Version 21.6.0: Mon Aug 22 20:19:52 PDT 2022; root:xnu-8020.140.49~2/RELEASE_ARM64_T6000
machine          : arm64
processor        : arm
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 1.5.0
numpy            : 1.23.3
pytz             : 2022.2.1
dateutil         : 2.8.2
setuptools       : 63.4.3
pip              : 22.2.2
Cython           : None
pytest           : 7.1.3
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.6.0
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.9.1
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None
tzdata           : None

### PR DESCRIPTION
**Pull request by AdamOrmondroyd**
- [x] closes #48884
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Review by phofl — CHANGES_REQUESTED** [review]
Hi, thx for this. Could you please add tests?

**Comment by Ormorod** [comment]
I've created a test where a df and set of weights are made, and one value in each column of df is set to NaN. I then make a copy of the df without the NaNs, and copies of weights with the corresponding weights dropped. It then checks that histograms made using each pair have the same heights.

File: pandas/tests/plotting/test_hist_method.py
Comment (mroeschke) [inline pandas/tests/plotting/test_hist_method.py:575]: Can you just initialize `no_nan_df`, `df`,  `no_nan_weights` with the `NaN`s manually? It will be easier for maintainers to follow in the future.

File: pandas/tests/plotting/test_hist_method.py
Comment (mroeschke) [inline pandas/tests/plotting/test_hist_method.py:597]:
```suggestion
        assert all(h0 == h1 for h0, h1 in zip(heights, no_nan_heights))
```

File: pandas/tests/plotting/test_hist_method.py
Comment (Ormorod) [reply pandas/tests/plotting/test_hist_method.py:575]: As opposed to choosing the NaN locations in each column randomly?

Comment (mroeschke) [reply pandas/tests/plotting/test_hist_method.py:575]: > As opposed to choosing the NaN locations in each column randomly?

Correct.

Comment (mroeschke) [reply pandas/tests/plotting/test_hist_method.py:575]: Thanks, but by manually, I was referring to that these objects should be constructed without and of the looping

```
df  = pd.DataFrame(...)
no_nan_df = DataFrame(...)
no_nan_weights = np.array (...)
```

Comment (Ormorod) [reply pandas/tests/plotting/test_hist_method.py:580]: I don't see how to produce the elements of `no_nan_df` or `no_nan_weights` without looping, unless you mean to manually specify every element?

`no_nan_df` contains the same elements as `df`, but removing the NaNs from each column. `no_nan_weights` then contains the elements of `weights` repeated 4 times, but each column omitting the weight corresponding to the NaN in `df`. This is so that the histograms produced by `df` and `weights` will be the same as `no_nan_df` and `no_nan_weights`.

File: pandas/plotting/_matplotlib/hist.py
Comment (mroeschke) [inline pandas/plotting/_matplotlib/hist.py:154]: Curious, why can't this block of code be in the the `if` statement above?

Comment (Ormorod) [reply pandas/plotting/_matplotlib/hist.py:151]: it can, if the `if` above is split in two. Just trying to change as few lines as possible :)

File: doc/source/whatsnew/v1.6.0.rst
Comment (mroeschke) [inline doc/source/whatsnew/v1.6.0.rst:246]: Could you write this in terms of the public api, `plot.hist`, as well as adding ```(:issue:`48884`)``` to the end?

**Review by mroeschke — APPROVED** [review]
Looks okay to me. @phofl merge when ready.

File: pandas/tests/plotting/test_hist_method.py
Comment (phofl) [inline pandas/tests/plotting/test_hist_method.py:568]: Can you clarify test name?

Comment (Ormorod) [reply pandas/tests/plotting/test_hist_method.py:568]: I've changed it to `test_hist_with_nans_and_weights`, is this clear enough?

File: pandas/plotting/_matplotlib/hist.py
Comment (phofl) [reply pandas/plotting/_matplotlib/hist.py:153]: This one is still open, otherwise lgtm

Comment (mroeschke) [reply pandas/plotting/_matplotlib/hist.py:153]: @Ormorod doesn't look like this was addressed. Can `i` be out of bounds?

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Comment by Ormorod** [comment]
> Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

Will merge main now. I addressed the comments above but haven't had a response.

**Comment by Ormorod** [comment]
It also occurs to me that `weights` itself might work better as a dict or DataFrame if it has more than a single column, but I think this is beyond the scope of this PR

File: pandas/plotting/_matplotlib/hist.py
Comment (mroeschke) [inline pandas/plotting/_matplotlib/hist.py:158]: Is there a tests that hits this exception?

Also we usually re-raise an exception from the original one (`raise ValueError(...) from err`)

**Review by mroeschke — APPROVED** [review]
LGTM merge when ready @phofl

**Review by phofl — APPROVED** [review]

**Comment by phofl** [comment]
Thx @Ormorod

**Pull request merged**
